src/ol/source/GeoTIFF.js

```diff
@@ -0,0 +1,462 @@
+        );
+        const message = `Resolution mismatch for source ${sourceIndex}, got [${scaledSourceResolutions}] but expected [${resolutions}]`;
+        assertEqual(
+          resolutions.slice(minZoom, resolutions.length),
+          scaledSourceResolutions,
+          0.005,
+          message
+        );
+      }
+
+      if (!tileSizes) {
+        tileSizes = sourceTileSizes;
+      } else {
+        assertEqual(
+          tileSizes.slice(minZoom, tileSizes.length),
+          sourceTileSizes,
+          0,
+          `Tile size mismatch for source ${sourceIndex}`
+        );
+      }
+
+      this.sourceImagery_[sourceIndex] = images.reverse();
+    }
+
+    for (let i = 0, ii = this.sourceImagery_.length; i < ii; ++i) {
+      const sourceImagery = this.sourceImagery_[i];
+      while (sourceImagery.length < resolutions.length) {
+        sourceImagery.unshift(undefined);
+      }
+    }
+
+    if (!this.getProjection()) {
+      const firstImage = sources[0][0];
+      if (firstImage.geoKeys) {
+        const code =
+          firstImage.geoKeys.ProjectedCSTypeGeoKey ||
+          firstImage.geoKeys.GeographicTypeGeoKey;
+        if (code) {
+          this.projection = getProjection(`EPSG:${code}`);
+        }
+      }
+    }
+
+    this.samplesPerPixel_ = samplesPerPixel;
+    const sourceInfo = this.sourceInfo_;
+    for (let sourceIndex = 0; sourceIndex < sourceCount; ++sourceIndex) {
+      if (sourceInfo[sourceIndex].nodata !== undefined) {
+        this.addAlpha_ = true;
+        break;
+      }
+    }
+    const additionalBands = this.addAlpha_ ? 1 : 0;
+    this.bandCount =
+      samplesPerPixel.reduce((accumulator, value) => {
+        accumulator += value;
+        return accumulator;
+      }, 0) + additionalBands;
+
+    const tileGrid = new TileGrid({
+      extent: extent,
+      minZoom: minZoom,
+      origin: origin,
+      resolutions: resolutions,
+      tileSizes: tileSizes,
+    });
+
+    this.tileGrid = tileGrid;
+
+    this.setLoader(this.loadTile_.bind(this));
+    this.setState(State.READY);
+  }
+
+  loadTile_(z, x, y) {
+    const size = toSize(this.tileGrid.getTileSize(z));
+
+    const sourceCount = this.sourceImagery_.length;
+    const requests = new Array(sourceCount);
+    const addAlpha = this.addAlpha_;
+    const bandCount = this.bandCount;
+    const samplesPerPixel = this.samplesPerPixel_;
+    const sourceInfo = this.sourceInfo_;
+    for (let sourceIndex = 0; sourceIndex < sourceCount; ++sourceIndex) {
+      const source = sourceInfo[sourceIndex];
+      const resolutionFactor = this.resolutionFactors_[sourceIndex];
+      const pixelBounds = [
+        Math.round(x * (size[0] * resolutionFactor)),
+        Math.round(y * (size[1] * resolutionFactor)),
+        Math.round((x + 1) * (size[0] * resolutionFactor)),
+        Math.round((y + 1) * (size[1] * resolutionFactor)),
+      ];
+      const image = this.sourceImagery_[sourceIndex][z];
+      requests[sourceIndex] = image.readRasters({
```

Comment:
Reading raw rasters is a good start. I think we could also somehow leverage the `readRGB` function (https://geotiffjs.github.io/geotiff.js/GeoTIFFImage.html#readRGB) which also translates color spaces such as YCbCr to RGB.

Most of the parameters are the same as `readRasters` anyhow.

Reply:
Like @tschaub already said, the problem with `readRGB()` is interleaving. @constantinius, what if `readRGB()` would also have an option to return non-interleaved data? The reason why we do not request interleaved data is because we shuffle bands and normalize values, which would add unnecessary overhead with interleaved data.

Reply:
I see, this is a problem as the whole RGB transformation is inherently interleaved. In that case, I need to think of something in geotiff.js.